tests/ui/missing_transmute_annotations.stderr

```diff
@@ -0,0 +1,40 @@
+error: transmute used without annotations
+  --> $DIR/missing_transmute_annotations.rs:11:32
+   |
+LL |         let x: i32 = std::mem::transmute([1u16, 2u16]);
+   |                                ^^^^^^^^^ help: consider adding missing annotations: `transmute::<[u16; 2], i32>`
+   |
+   = note: `-D clippy::missing-transmute-annotations` implied by `-D warnings`
+   = help: to override `-D warnings` add `#[allow(clippy::missing_transmute_annotations)]`
+
+error: transmute used without annotations
+  --> $DIR/missing_transmute_annotations.rs:13:32
+   |
+LL |         let x: i32 = std::mem::transmute::<_, _>([1u16, 2u16]);
+   |                                ^^^^^^^^^^^^^^^^^ help: consider adding missing annotations: `transmute::<[u16; 2], i32>`
+
+error: transmute used without annotations
+  --> $DIR/missing_transmute_annotations.rs:15:27
+   |
+LL |         let x = std::mem::transmute::<_, i32>([1u16, 2u16]);
+   |                           ^^^^^^^^^^^^^^^^^^^ help: consider adding missing annotations: `transmute::<[u16; 2], i32>`
+
+error: transmute used without annotations
+  --> $DIR/missing_transmute_annotations.rs:17:32
+   |
+LL |         let x: i32 = std::mem::transmute::<[u16; 2], _>([1u16, 2u16]);
```

Comment:
`_` really shouldn't be linted, you're explicitly saying you want it to be inferred

Reply:
Indeed, I implemented that.

Reply:
Isn't this still linted?
https://github.com/rust-lang/rust-clippy/blob/235c9f397f36f3c89d7122851776d8421ccf8fcc/tests/ui/missing_transmute_annotations.stderr#L22-L26

Reply:
Unless I missed something, there is no `let` binding with type annotation on this line.

Reply:
Ah, I might've misunderstood @Centri3's comment. The way I understood it was that it should avoid linting entirely when `_` is specified no matter what, whereas the current impl allows inferred types when there is an enclosing type-annotated binding.

If I did understand it correctly, I'm not sure if it's worth special casing explicit `_` in the generic arguments and treating them differently like that (which it's currently not doing anyway). All the same caveats and inference arguments still apply when writing `transmute::<_, _>()` as opposed to just `transmute()`.

Thinking about it, I do like the current heuristics (checking if the parent node is a local variable with a type annotation).
It might make sense to actually extend this to also check if this is the last expression in the function and not lint then? I mentioned this in the other comment about the lint category and IMHO it would rule out a lot of noise.
I did a lintcheck run and almost all of the hits were small functions where it's just a transmute and the target type is already annotated in the function signature.

We could also have a configuration for a 'stricter' mode, to not even allow these cases, which I think is a good idea if we're gonna leave it as a warn-by-default lint. WDYT?

Reply:
I wasn't thinking about the `let` at all, I think it shouldn't lint it with `transmute::<_, _>()`. I can't think of any reasons to actually want it inferred, but better safe than sorry. A stricter mode should definitely be added if it's warn-by-default, and maybe if it's allow-by-default too

Reply:
> It might make sense to actually extend this to also check if this is the last expression in the function and not lint then? 

We already discussed about it in this PR and it was actually one of the motivations for this PR: changing the returned type of a function can alter the `transmute` returned value and that is likely not what the user would want.

@Centri3: I don't think it's a good idea. Having `transmute::<_, _>(` is the same as having `transmute(`. It brings no additional information and should be linted against.

Reply:
It does give more information, though: you want it to be inferred. Why's that? No idea. If this is allow-by-default clearly somebody wants to apply the lint's suggestion, and won't do this just to silence clippy.

And playing the devil's advocate, you can't always specify a generic, especially so if it isn't even generic. Say in macro generated code, or just generated code in general. This is absolutely reckless but if you're using transmute, you hopefully know what you're doing anyway, and any preconditions are upheld elsewhere. At the end of the day you can just `#[allow]` it, but that wouldn't be linted if you switch to the stricter mode (say you're phasing this out of your codebase, or something, I dunno)

Reply:
Is there a way to check how many crates would be impacted by this lint and advise from that? Imo, if we don't emit the lint if `_` is used of any of the generics, then the lint loses a big part of its use.

Reply:
Try running lintcheck and only lint if `_` is used, I suppose. I don't really mind if this is denied by default so that those cases (where somebody uses `_` erroneously) are still linted. This could be tied to a configuration option

I do agree with you that those cases are important, now that I think about it.

Reply:
> We already discussed about it in this PR and it was actually one of the motivations for this PR: changing the returned type of a function can alter the transmute returned value and that is likely not what the user would want.

For long functions where you might not be able to immediately spot the type annotation in the signature, I agree. But for functions where the transmute call is the only expression in the function and there are no other statements, I'm not so sure after thinking about it a bit more.
For example, it seems excessive to have to write
```rs
fn i32_to_ne_bytes(val: i32) -> [u8; 4] {
  unsafe { std::mem::transmute::<i32, [u8; 4]>(val) }
}
```
Arguably, this is just as local as a `let` binding with type annotations

Reply:
Whatever the length of the function, it's easy to forget to update both the returned type of the function and eventually check the `transmute`.

Reply:
I'm still not entirely sure here. I opened a Zulip stream for this, just to get some more opinions: https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/missing_transmute_annotations

Reply:
Looks like the "majority" on the zulip thread was of the opinion that we should lint this, so let's go with that, then.

I also hadn't considered that people might fold functions in their IDEs such that the transmute is not as apparent, so that's another point for why we should lint this anyway